[BUGFIX beta] Ensure that `this._super` is called by Components.

rwjblue opened the pull request:
When implementing `init` in a component subclass you must call `this._super(...arguments)`.  If you do not you will get an error.  Unfortunately, that error is **very** hard to understand and is completely unrelated to code that you have written.

This change adds a small flag (via symbol to avoid exposing this publicly) that we can use to ensure that `_super` was called properly.

---

After these changes, given the following:

``` javascript
import Ember from 'ember';

export default Ember.Component.extend({
  init() {
    this.doThings();
  }
});
```

Will throw an error that explains that you must call `_super`:

```
You must call `this._super(...arguments);` when implementing `init`
in a component. Please update ${this} to call `this._super` from `init`.
```

/cc @ef4

rwjblue committed:
[BUGFIX beta] Ensure that `this._super` is called by Components.

When implementing `init` in a component subclass you must call
`this._super(...arguments)`.  If you do not you will get an error.
Unfortunately, that error is **very** hard to understand and is
completely unrelated to code that you have written.

This change adds a small flag (via symbol to avoid exposing this
publicly) that we can use to ensure that `_super` was called
properly.

---

After these changes, given the following:

```javascript
import Ember from 'ember';

export default Ember.Component.extend({
  init() {
    this.doThings();
  }
});
```

Will throw an error that explains that you must call `_super`:

```
You must call `this._super(...arguments);` when implementing `init`
in a component. Please update ${this} to call `this._super` from `init`.
```

ef4 commented on packages/ember-views/lib/mixins/view_support.js:
```diff
@@ -7,6 +7,9 @@ import { guidFor } from 'ember-metal/utils';
 import { POST_INIT } from 'ember-runtime/system/core_object';
+import { symbol } from 'ember-metal/utils';
+
+let INIT_WAS_CALLED = symbol('INIT_WAS_CALLED');
```
Opportunity for `const`.

rwjblue replied:
Good point, updated!

stefanpenner commented on packages/ember-views/lib/mixins/view_support.js:
```diff
@@ -610,6 +613,7 @@ export default Mixin.create({
     this.scheduledRevalidation = false;
```
although not part of this commit, should this assignment to `this` be before the `init` ?

rwjblue replied:
I'm not sure I follow what you mean.  This assignment is happening just before `this._super` is called, do you want to move it afterwards or something else?

stefanpenner replied:
yes, as in ES6 you can't touch `this` before calling super, we should slowly move our internals to this.

rwjblue replied:
Updated to call `this._super` first (before using anything on `this`).

stefanpenner commented:
LGTM

rwjblue committed:
[BUGFIX beta] Call `this._super` before touching `this`.

You should generally always call `this._super(...arguments);` before
touching `this` in the constructor. This requirement is actually
enforced by ES2015 classes...
